Sometimes the sensing module doesn't publish concatenated pointcloud

### Checklist

- [X] I've read the [contribution guidelines](https://github.com/autowarefoundation/autoware/blob/main/CONTRIBUTING.md).
- [X] I've searched other issues and no duplicate issues were found.
- [X] I'm convinced that this is not my fault but a bug.

### Description

sometimes the sensing module cannot launch concatenated and cropbox filter with the following error
```
[ERROR] [launch_ros.actions.load_composable_nodes]: Failed to load node 'concatenate_data' of type 'pointcloud_preprocessor::PointCloudConcatenateDataSynchronizerComponent' in container 'pointcloud_container': Failed to find class with the requested plugin name.
[ERROR] [launch_ros.actions.load_composable_nodes]: Failed to load node 'crop_box_filter' of type 'pointcloud_preprocessor::CropBoxFilterComponent' in container 'pointcloud_container': Failed to find class with the requested plugin name.
```

### Expected behavior

launch concatenated and cropbox filter without error

### Actual behavior

launch concatenated and cropbox filter without error

### Steps to reproduce

play rosbag replay simulator in tutorials

### Versions

_No response_

### Possible causes

_No response_

### Additional context

_No response_

### Temporary fix

- [ ] #1797

I can't reproduce the issue, but from the log my colleague shared, I guess some components which are supposed to exist aren't found as below.
```
[component_container_mt-1] [ERROR] [1658719110.290187926] [pointcloud_container]: Failed to find class with the requested plugin name 'pointcloud_preprocessor::CropBoxFilterComponent' in the loaded library
[ERROR] [launch_ros.actions.load_composable_nodes]: Failed to load node 'crop_box_filter' of type 'pointcloud_preprocessor::CropBoxFilterComponent' in container 'pointcloud_container': Failed to find class with the requested plugin name.
[INFO] [launch_ros.actions.load_composable_nodes]: Loaded node '/perception/object_recognition/detection/clustering/euclidean_cluster' in container '/perception/object_recognition/detection/clustering/euclidean_cluster_container'
[ERROR] [launch_ros.actions.load_composable_nodes]: Failed to load node 'voxel_grid_downsample_filter' of type 'pointcloud_preprocessor::VoxelGridDownsampleFilterComponent' in container 'pointcloud_container': Failed to find class with the requested plugin name.
[component_container_mt-1] [ERROR] [1658719110.292822374] [pointcloud_container]: Failed to find class with the requested plugin name
```

During debugging I noticed that launch sequence is not a root cause. For whatever reason, when the failure occurs when   `libcompare_map_segmentation.so ` is loaded before  `libpointcloud_preprocessor_filter.so`

```
[INFO] [component_container_mt-1]: process started with pid [1359167]
[component_container_mt-1] [INFO] [1664351258.345340026] [pointcloud_container]: Load Library: /home/lukasz/projects/autoware_official_main/autoware/install/compare_map_segmentation/lib/libcompare_map_segmentation.so
[component_container_mt-1] [INFO] [1664351266.708360676] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<compare_map_segmentation::CompareElevationMapFilterComponent>
[component_container_mt-1] [INFO] [1664351266.708518922] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<compare_map_segmentation::DistanceBasedCompareMapFilterComponent>
[component_container_mt-1] [INFO] [1664351266.708557455] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<compare_map_segmentation::VoxelBasedApproximateCompareMapFilterComponent>
[component_container_mt-1] [INFO] [1664351266.708591027] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<compare_map_segmentation::VoxelBasedCompareMapFilterComponent>
[component_container_mt-1] [INFO] [1664351266.708623150] [pointcloud_container]: Instantiate class: rclcpp_components::NodeFactoryTemplate<compare_map_segmentation::VoxelBasedCompareMapFilterComponent>
[component_container_mt-1] [INFO] [1664351267.132734306] [perception.object_recognition.detection.voxel_based_compare_map_filter]: Filter (as Component) successfully created with the following parameters:
[component_container_mt-1]  - approximate_sync : false
[component_container_mt-1]  - use_indices      : false
[component_container_mt-1]  - latched_indices  : false
[component_container_mt-1]  - max_queue_size   : 5

[component_container_mt-1] [INFO] [1664351267.385921052] [pointcloud_container]: Load Library: /home/lukasz/projects/autoware_official_main/autoware/install/pointcloud_preprocessor/lib/libpointcloud_preprocessor_filter.so
[component_container_mt-1] [ERROR] [1664351267.430029960] [pointcloud_container]: Failed to find class with the requested plugin name 'pointcloud_preprocessor::PointCloudConcatenateDataSynchronizerComponent' in the loaded library
[component_container_mt-1] [ERROR] [1664351267.436821745] [pointcloud_container]: Failed to find class with the requested plugin name 'pointcloud_preprocessor::CropBoxFilterComponent' in the loaded library

[component_container_mt-1] [INFO] [1664351267.774941134] [pointcloud_container]: Load Library: /home/lukasz/projects/autoware_official_main/autoware/install/ground_segmentation/lib/libground_segmentation.so
[component_container_mt-1] [INFO] [1664351267.826568706] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<ground_segmentation::RANSACGroundFilterComponent>
[component_container_mt-1] [INFO] [1664351267.826816182] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<ground_segmentation::RayGroundFilterComponent>
[component_container_mt-1] [INFO] [1664351267.826862295] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<ground_segmentation::ScanGroundFilterComponent>
[component_container_mt-1] [INFO] [1664351267.826898226] [pointcloud_container]: Instantiate class: rclcpp_components::NodeFactoryTemplate<ground_segmentation::ScanGroundFilterComponent>
[component_container_mt-1] [INFO] [1664351268.760633951] [perception.obstacle_segmentation.common_ground_filter]: Filter (as Component) successfully created with the following parameters:
[component_container_mt-1]  - approximate_sync : false
[component_container_mt-1]  - use_indices      : false
[component_container_mt-1]  - latched_indices  : false
[component_container_mt-1]  - max_queue_size   : 5
[component_container_mt-1] [INFO] [1664351269.408068814] [pointcloud_container]: Load Library: /home/lukasz/projects/autoware_official_main/autoware/install/occupancy_grid_map_outlier_filter/lib/liboccupancy_grid_map_outlier_filter.so
[component_container_mt-1] [INFO] [1664351269.419959183] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<occupancy_grid_map_outlier_filter::OccupancyGridMapOutlierFilterComponent>
[component_container_mt-1] [INFO] [1664351269.420117594] [pointcloud_container]: Instantiate class: rclcpp_components::NodeFactoryTemplate<occupancy_grid_map_outlier_filter::OccupancyGridMapOutlierFilterComponent>
[component_container_mt-1] [WARN] [1664351283.256087367] [perception.obstacle_segmentation.occupancy_grid_map_outlier_filter]: New subscription discovered on topic '/perception/obstacle_segmentation/point cloud, requesting incompatible QoS. No messages will be sent to it. Last incompatible policy: RELIABILITY_QOS_POLICY
```

When `libpointcloud_preprocessor_filter.so` is loaded before  `libcompare_map_segmentation.so` then everything is ok.
```
[INFO] [component_container_mt-1]: process started with pid [1345975]

[component_container_mt-1] [INFO] [1664350962.225026369] [pointcloud_container]: Load Library: /home/lukasz/projects/autoware_official_main/autoware/install/pointcloud_preprocessor/lib/libpointcloud_preprocessor_filter.so
[component_container_mt-1] [INFO] [1664350969.753593880] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<pointcloud_preprocessor::ApproximateDownsampleFilterComponent>
[component_container_mt-1] [INFO] [1664350969.761015939] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<pointcloud_preprocessor::BlockageDiagComponent>
[component_container_mt-1] [INFO] [1664350969.761086550] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<pointcloud_preprocessor::CropBoxFilterComponent>
[component_container_mt-1] [INFO] [1664350969.761123733] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<pointcloud_preprocessor::DistortionCorrectorComponent>
[component_container_mt-1] [INFO] [1664350969.761157587] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<pointcloud_preprocessor::DualReturnOutlierFilterComponent>
[component_container_mt-1] [INFO] [1664350969.761191740] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<pointcloud_preprocessor::Lanelet2MapFilterComponent>
[component_container_mt-1] [INFO] [1664350969.761225317] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<pointcloud_preprocessor::PassThroughFilterComponent>
[component_container_mt-1] [INFO] [1664350969.761258883] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<pointcloud_preprocessor::PassThroughFilterUInt16Component>
[component_container_mt-1] [INFO] [1664350969.761293253] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<pointcloud_preprocessor::PointCloudConcatenateDataSynchronizerComponent>
[component_container_mt-1] [INFO] [1664350969.761327833] [pointcloud_container]: Instantiate class: rclcpp_components::NodeFactoryTemplate<pointcloud_preprocessor::PointCloudConcatenateDataSynchronizerComponent>


[component_container_mt-1] [INFO] [1664350970.970393159] [sensing.lidar.concatenate_data]: Subscribing to 3 user given topics as inputs:
[component_container_mt-1] [INFO] [1664350970.977884115] [sensing.lidar.concatenate_data]:  - /sensing/lidar/top/outlier_filtered/pointcloud
[component_container_mt-1] [INFO] [1664350970.979455088] [sensing.lidar.concatenate_data]:  - /sensing/lidar/left/outlier_filtered/pointcloud
[component_container_mt-1] [INFO] [1664350970.979509090] [sensing.lidar.concatenate_data]:  - /sensing/lidar/right/outlier_filtered/pointcloud

[component_container_mt-1] [INFO] [1664350971.151433649] [pointcloud_container]: Load Library: /home/lukasz/projects/autoware_official_main/autoware/install/compare_map_segmentation/lib/libcompare_map_segmentation.so
[component_container_mt-1] [INFO] [1664350971.579233393] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<compare_map_segmentation::CompareElevationMapFilterComponent>
[component_container_mt-1] [INFO] [1664350971.579409082] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<compare_map_segmentation::DistanceBasedCompareMapFilterComponent>
[component_container_mt-1] [INFO] [1664350971.581189423] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<compare_map_segmentation::VoxelBasedApproximateCompareMapFilterComponent>
[component_container_mt-1] [INFO] [1664350971.582902671] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<compare_map_segmentation::VoxelBasedCompareMapFilterComponent>
[component_container_mt-1] [INFO] [1664350971.582951453] [pointcloud_container]: Instantiate class: rclcpp_components::NodeFactoryTemplate<compare_map_segmentation::VoxelBasedCompareMapFilterComponent>
[component_container_mt-1] [INFO] [1664350972.569075844] [perception.object_recognition.detection.voxel_based_compare_map_filter]: Filter (as Component) successfully created with the following parameters:
[component_container_mt-1]  - approximate_sync : false
[component_container_mt-1]  - use_indices      : false
[component_container_mt-1]  - latched_indices  : false
[component_container_mt-1]  - max_queue_size   : 5
[component_container_mt-1] [INFO] [1664350972.898912106] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<pointcloud_preprocessor::ApproximateDownsampleFilterComponent>
[component_container_mt-1] [INFO] [1664350972.899934429] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<pointcloud_preprocessor::BlockageDiagComponent>
[component_container_mt-1] [INFO] [1664350972.899980698] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<pointcloud_preprocessor::CropBoxFilterComponent>
[component_container_mt-1] [INFO] [1664350972.901752718] [pointcloud_container]: Instantiate class: rclcpp_components::NodeFactoryTemplate<pointcloud_preprocessor::CropBoxFilterComponent>
[component_container_mt-1] [INFO] [1664350973.984691344] [perception.obstacle_segmentation.crop_box_filter]: Filter (as Component) successfully created with the following parameters:
[component_container_mt-1]  - approximate_sync : false
[component_container_mt-1]  - use_indices      : false
[component_container_mt-1]  - latched_indices  : false
[component_container_mt-1]  - max_queue_size   : 5
[component_container_mt-1] [INFO] [1664350974.160625220] [pointcloud_container]: Load Library: /home/lukasz/projects/autoware_official_main/autoware/install/ground_segmentation/lib/libground_segmentation.so
[component_container_mt-1] [INFO] [1664350974.334012500] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<ground_segmentation::RANSACGroundFilterComponent>
[component_container_mt-1] [INFO] [1664350974.334175819] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<ground_segmentation::RayGroundFilterComponent>
[component_container_mt-1] [INFO] [1664350974.334214552] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<ground_segmentation::ScanGroundFilterComponent>
[component_container_mt-1] [INFO] [1664350974.336807913] [pointcloud_container]: Instantiate class: rclcpp_components::NodeFactoryTemplate<ground_segmentation::ScanGroundFilterComponent>
[component_container_mt-1] [INFO] [1664350975.661686566] [perception.obstacle_segmentation.common_ground_filter]: Filter (as Component) successfully created with the following parameters:
[component_container_mt-1]  - approximate_sync : false
[component_container_mt-1]  - use_indices      : false
[component_container_mt-1]  - latched_indices  : false
[component_container_mt-1]  - max_queue_size   : 5
[component_container_mt-1] [INFO] [1664350976.050537905] [pointcloud_container]: Load Library: /home/lukasz/projects/autoware_official_main/autoware/install/occupancy_grid_map_outlier_filter/lib/liboccupancy_grid_map_outlier_filter.so
[component_container_mt-1] [INFO] [1664350976.092318710] [pointcloud_container]: Found class: rclcpp_components::NodeFactoryTemplate<occupancy_grid_map_outlier_filter::OccupancyGridMapOutlierFilterComponent>
[component_container_mt-1] [INFO] [1664350976.092452991] [pointcloud_container]: Instantiate class: rclcpp_components::NodeFactoryTemplate<occupancy_grid_map_outlier_filter::OccupancyGridMapOutlierFilterComponent>
[component_container_mt-1] [WARN] [1664350983.999561958] [sensing.lidar.concatenate_data]: New subscription discovered on topic '/sensing/lidar/concatenated/point cloud, requesting incompatible QoS. No messages will be sent to it. Last incompatible policy: RELIABILITY_QOS_POLICY
[component_container_mt-1] [WARN] [1664350984.264477043] [perception.obstacle_segmentation.occupancy_grid_map_outlier_filter]: New subscription discovered on topic '/perception/obstacle_segmentation/point cloud, requesting incompatible QoS. No messages will be sent to it. Last incompatible policy: RELIABILITY_QOS_POLICY
```

The main question is what determines the order of loading .so modules? 
Unfortunately, I don't know the answer.

--- a/perception/compare_map_segmentation/CMakeLists.txt
+++ b/perception/compare_map_segmentation/CMakeLists.txt
@@ -21,7 +21,7 @@ include_directories(
   ${GRID_MAP_INCLUDE_DIR}
 )
 
-ament_auto_add_library(compare_map_segmentation SHARED
+add_library(compare_map_segmentation SHARED
   src/distance_based_compare_map_filter_nodelet.cpp
   src/voxel_based_approximate_compare_map_filter_nodelet.cpp
   src/voxel_based_compare_map_filter_nodelet.cpp
@@ -30,9 +30,20 @@ ament_auto_add_library(compare_map_segmentation SHARED
 )
 
 target_link_libraries(compare_map_segmentation
+  pointcloud_preprocessor::pointcloud_preprocessor_filter_base
   ${Boost_LIBRARIES}
   ${OpenCV_LIBRARIES}
   ${PCL_LIBRARIES}
+)
+
+ament_target_dependencies(compare_map_segmentation
+  grid_map_pcl
+  grid_map_ros
+  pcl_conversions
+  rclcpp
+  rclcpp_components
+  sensor_msgs
+  tier4_autoware_utils
 )
 
 if(OPENMP_FOUND)
@@ -67,6 +78,13 @@ rclcpp_components_register_node(compare_map_segmentation
 rclcpp_components_register_node(compare_map_segmentation
   PLUGIN "compare_map_segmentation::CompareElevationMapFilterComponent"
   EXECUTABLE compare_elevation_map_filter_node)
+
+install(
+  TARGETS compare_map_segmentation
+  ARCHIVE DESTINATION lib
+  LIBRARY DESTINATION lib
+  RUNTIME DESTINATION bin
+)
 
 ament_auto_package(INSTALL_TO_SHARE
   launch

--- a/sensing/pointcloud_preprocessor/CMakeLists.txt
+++ b/sensing/pointcloud_preprocessor/CMakeLists.txt
@@ -21,8 +21,25 @@ include_directories(
   ${GRID_MAP_INCLUDE_DIR}
 )
 
-ament_auto_add_library(pointcloud_preprocessor_filter SHARED
+add_library(pointcloud_preprocessor_filter_base SHARED
   src/filter.cpp
+)
+
+target_include_directories(pointcloud_preprocessor_filter_base PUBLIC
+  "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
+  "$<INSTALL_INTERFACE:include/${PROJECT_NAME}>"
+)
+
+ament_target_dependencies(pointcloud_preprocessor_filter_base
+  message_filters
+  pcl_conversions
+  rclcpp
+  sensor_msgs
+  tf2_ros
+  tier4_autoware_utils
+)
+
+ament_auto_add_library(pointcloud_preprocessor_filter SHARED
   src/utility/utilities.cpp
   src/concatenate_data/concatenate_data_nodelet.cpp
   src/crop_box_filter/crop_box_filter_nodelet.cpp
@@ -45,6 +62,7 @@ ament_auto_add_library(pointcloud_preprocessor_filter SHARED
 )
 
 target_link_libraries(pointcloud_preprocessor_filter
+  pointcloud_preprocessor_filter_base
   ${Boost_LIBRARIES}
   ${OpenCV_LIBRARIES}
   ${PCL_LIBRARIES}
@@ -146,6 +164,20 @@ target_link_libraries(polygon_remover_node gmp CGAL CGAL::CGAL CGAL::CGAL_Core)
 rclcpp_components_register_node(pointcloud_preprocessor_filter
   PLUGIN "pointcloud_preprocessor::VectorMapInsideAreaFilterComponent"
   EXECUTABLE vector_map_inside_area_filter_node)
+
+install(
+  TARGETS pointcloud_preprocessor_filter_base EXPORT export_${PROJECT_NAME}
+  ARCHIVE DESTINATION lib
+  LIBRARY DESTINATION lib
+  RUNTIME DESTINATION bin
+)
+
+install(
+  DIRECTORY include/
+  DESTINATION include/${PROJECT_NAME}
+)
+
+ament_export_targets(export_${PROJECT_NAME})
 
 ament_auto_package(INSTALL_TO_SHARE
   launch

